delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/ResourceResolution.scala

```diff
@@ -0,0 +1,162 @@
+import ch.epfl.bluebrain.nexus.delta.sdk.model.resolvers.{Resolver, ResolverRejection, ResourceResolutionReport}
+import ch.epfl.bluebrain.nexus.delta.sdk.model.search.SearchParams.ResolverSearchParams
+import ch.epfl.bluebrain.nexus.delta.sdk.model.search.{Pagination, ResultEntry}
+import ch.epfl.bluebrain.nexus.delta.sdk.model.{IdSegment, ResourceRef}
+import monix.bio.{IO, UIO}
+
+/**
+  * Resolution for a given type of resource
+  * @param acls            the acls module
+  * @param resolvers       the resolvers module
+  * @param readPermission  the permission required to get the given type of resource on a project
+  * @param fetchResource   how we can get a resource from a [[ResourceRef]]
+  * @param lensTypes       a lens to extract the types from the given resource
+  */
+final class ResourceResolution[Resource](
+    acls: Acls,
+    resolvers: Resolvers,
+    readPermission: Permission,
+    fetchResource: (ResourceRef, ProjectRef) => IO[FetchRejection, Resource]
+)(implicit lensTypes: Lens[Resource, Set[Iri]]) {
+
+  private val resolverSearchParams = ResolverSearchParams(deprecated = Some(false), filter = _ => true)
+
+  /**
+    * Attempts to resolve the resource against all resolvers of the given project by priority order,
+    * discard the report when it succeeds, raises it as an error if resolution fails
+    *
+    * @param ref        the resource reference
+    * @param projectRef the project reference
+    */
+  def resolve(ref: ResourceRef, projectRef: ProjectRef)(implicit
+      caller: Caller
+  ): IO[ResourceResolutionReport, Resource] =
+    resolveReport(ref, projectRef).flatMap { r =>
+      IO.fromOption(r._2, r._1)
+    }
+
+  /**
+    * Attempts to resolve the resource against all resolvers of the given project by priority order,
+    * discard the report when it succeeds, raises it as an error if resolution fails
+    *
+    * @param ref        the resource reference
+    * @param projectRef the project reference
+    */
+  def resolveReport(ref: ResourceRef, projectRef: ProjectRef)(implicit
+      caller: Caller
+  ): UIO[(ResourceResolutionReport, Option[Resource])] = {
+    val initial: (ResourceResolutionReport, Option[Resource]) = ResourceResolutionReport(List.empty) -> None
+    resolvers
+      .list(
+        projectRef,
+        Pagination.OnePage,
+        resolverSearchParams
+      )
+      .map(_.results.map { r: ResultEntry[ResolverResource] => r.source.value }.sortBy(_.priority.value))
+      .flatMap {
+        _.toList.foldLeftM(initial) { (previous, resolver) =>
+          previous match {
+            // A resolver was able to get the resource, we keep the result
+            case (report, result @ Some(_)) => IO.pure(report -> result)
+            // No resolution was successful yet, we carry on
+            case (report, None)             =>
+              resolveReport(ref, projectRef, resolver).map { case (resolverReport, result) =>
```

Comment:
Yes. In case of multiple crossProjectResolvers make sure you just query once the entire ACL collection. Probably you want to use this call one time: 
```scala
acls.list(AnyOrganizationAnyProject(withAncestors = true))
```
And afterwards filter for the self depending on the identities on each of the cross project resolvers.

Reply:
I chose to memoize the call so that we call it once for the first cross-project resolver and never when there is none